File: src/csharp/4_BehavioralPatterns/7_NullObject/NullObject.cs
using System;
using System.Dynamic;
using ImpromptuInterface;
using static System.Console;

namespace DotNetDesignPatternDemos.Behavioral.NullObject
{
  public interface ILog
  {
    void Info(string msg);
    void Warn(string msg);
  }

  class ConsoleLog : ILog
  {
    public void Info(string msg)
    {
      WriteLine(msg);
    }

    public void Warn(string msg)
    {
      WriteLine("WARNING: " + msg);
    }
  }

  public class BankAccount
  {
    private ILog log;
    private int balance;

    public BankAccount(ILog log)
    {
      this.log = log;
    }

    public void Deposit(int amount)
    {
      balance += amount;
      // check for null everywhere
      log?.Info($"Deposited ${amount}, balance is now {balance}");
    }

    public void Withdraw(int amount)
    {
      if (balance >= amount)
      {
        balance -= amount;
        log?.Info($"Withdrew ${amount}, we have ${balance} left");
      }
      else
      {
        log?.Warn($"Could not withdraw ${amount} because balance is only ${balance}");
      }
    }
  }

  public sealed class NullLog : ILog
  {
    public void Info(string msg)
    {
      
    }

    public void Warn(string msg)
    {
      
    }
  }

  public class Null<T> : DynamicObject where T:class
  {
    public static T Instance 
    {
      get
      {
        if (!typeof(T).IsInterface)
          throw new ArgumentException("I must be an interface type");

        return new Null<T>().ActLike<T>();
      }
    }

    public override bool TryInvokeMember(InvokeMemberBinder binder, object[] args, out object result)
    {
      result = Activator.CreateInstance(binder.ReturnType);
      return true;
    }

    private class Empty { }
  }

  public class Demo
  {
    static void Main()
    {
      //var log = new ConsoleLog();
      //ILog log = null;
      //var log = new NullLog();
      var log = Null<ILog>.Instance;
      var ba = new BankAccount(log);
      ba.Deposit(100);
      ba.Withdraw(200);
    }
  }
}

File: src/java/4_BehavioralPatterns/7_NullObject/NullObject.java
package com.activemesa.behavioral.nullobject;

interface Log
{
  void info(String msg);
  void warn(String msg);
}

class ConsoleLog implements Log
{

  @Override
  public void info(String msg)
  {
    System.out.println(msg);
  }

  @Override
  public void warn(String msg)
  {
    System.out.println("WARNING: " + msg);
  }
}

class BankAccount
{
  private Log log;
  private int balance;

  public BankAccount(Log log)
  {
    this.log = log;
  }

  public void deposit(int amount)
  {
    balance += amount;

    // check for null everywhere?
    if (log != null)
    {
      log.info("Deposited " + amount
        + ", balance is now " + balance);
    }
  }

  public void withdraw(int amount)
  {
    if (balance >= amount)
    {
      balance -= amount;
      if (log != null)
      {
        System.out.println("Withdrew " + amount
          + ", we have " + balance + " left");
      }
    }
    else {
      if (log != null)
      {
        System.out.println("Could not withdraw "
          + amount + " because balance is only " + balance);
      }
    }
  }
}

final class NullLog implements Log
{

  @Override
  public void info(String msg)
  {

  }

  @Override
  public void warn(String msg)
  {

  }
}

class NullObjectDemo
{
  public static void main(String[] args)
  {
    //ConsoleLog log = new ConsoleLog();
    //Log log = null;
    NullLog log = new NullLog();

    BankAccount ba = new BankAccount(log);
    ba.deposit(100);
    ba.withdraw(200);
  }
}